text/0000-modular-queries.md

```diff
@@ -0,0 +1,274 @@
+
+```
+DefineSubqueryDefn ::= DocComment? 'define' 'subquery' Id ('from' Ports)? ('into' Ports)? WithPartialParams? Subquery
+```
+
+```
+## Documentation Comment
+define subquery custom_subquery
+from input_stream_1, input_stream_2
+into output_stream
+with
+ param1 = "foo",
+ param2 = 42
+subquery
+ select event from input_stream_1 where event.name == args.param1 into output_stream;
+ select event from input_stream_2 where event.id == args.param2 into output_stream;
+end;
+```
+
+- The `with` clause can be used to pass in parameters when required.
+    - The values given to the parameters in the with clause here act as their default values.
+    - Parameters are accessible through `args` inside the subquery.
+- The `from` clause can be used to define input streams.
+    - If elided, `in` stream is attached.
+- The `into` clause can be used to define output streams.
+    - If elided, `out` and `err` streams are attached.  
+&nbsp;
+- **Note**
+    - Although there are currently no restrictions on sending events into the `input_stream` or reading from the `output_stream`, it is **not** recommended to do so.
+    - The streams named inside the `from` and `into` clauses are created inside the subquery implicitly. i.e. there is no need to `create stream output_stream` inside the subquery.
+
+## Creation
+
+![CreateSubqueryDefn](../img/rfcs/0000-modular-queries/CreateSubqueryDefn.png)
+
+```
+CreateSubqueryDefn ::= 'create' 'subquery' Id ( 'from' ModularId )? WithParams?
+```
+
+```
+# Short form
+create subquery custom_subquery;
+
+# Full form
+create subquery my_custom_subq from custom_subquery
+with
+ param1 = "bar"
+end;
+```
+
+- The short form can be used if you don’t need to give the subquery a custom id.
+    - The `id` from the subquery definition is used, ie `custom_subquery` in this case.
+- The full form allows you to give a custom id to the subquery.
+- The `with` clause can be used with either form to specify some or all of the parameters. Default values from the definition will be used for those left unspecified.
+
+## Use
+
+```
+select event from in into my_custom_subq/input_stream_1;
+select event from in into my_custom_subq/input_stream_2;
+select event from my_custom_subq/output_stream into out;
+```
+
+- We need to explicitly specify the ports with `my_custom_subq`  here because the subquery is not using the default `in`, `out` or `err` ports.
+
+## Example 1
+
+```
+mod library with
+ mod utils with
+
+   define script mark_malformed
+   script
+     emit {
+     "event": event,
+     "status": "malformed"
+   } => "invalid";
+   end;
+
+   define subquery select_minage
+     with
+       age = 18 # Parameter with default value of `18`
+     subquery
+       select event from in where in.age >= args.age into out;
+   end;
+
+ end;
+
+ define subquery select_valid_people
+ with
+   age = 21,
+   placeholder_name = "NA"
+ subquery
+   use utils;
+   create script mark_malformed;
```

Comment:
This tripped me up upon first reading, because i missed the `use utils;` line above.
It wouldnt be possible to just reference those scripts as `create script utils::mark_malformed`, right? A subquery should be self-contained. A `use` statement effectively moves the definitions inside this subquery, a simple reference wouldnt work here, and i think is illegal in trickle right now anyways, am i right?

Reply:
> It wouldnt be possible to just reference those scripts as `create script utils::mark_malformed`, right?

That would be possible.
It is allowed to `create` from a definition outside a subquery by _either_ `use`ing the module containing it or by giving the qualified path.

However, It is **not** allowed to pass in references as parameters via the `with` clause.